File: Qsort_3w/src/Qsort_3w.py
import sys, functools
import random

def partition3(a, l, r):
    #write your code here
    x = a[l]
    i = l+1
    L = []
    C = [x]
    R = []
    lt = l
    gt = r+1
    for i in range(l+1, r+1):
      if a[i] < x:
        L.append(a[i])
        lt += 1
      elif a[i] == x:
        C.append(a[i])
      else:
        R.append(a[i])
        gt -= 1
    #print x, L, C, R
    a[l:r+1] = L + C + R
    #print a[l:r+1], lt, gt
    return lt-1, gt

def partition2(a, l, r):
    x = a[l]
    j = l
    for i in range(l + 1, r + 1):
        if a[i] <= x:
            j += 1
            a[i], a[j] = a[j], a[i]
    a[l], a[j] = a[j], a[l]
    return j


def randomized_quick_sort(a, l, r):
    if l >= r:
        return
    k = random.randint(l, r)
    a[l], a[k] = a[k], a[l]
    #m = partition2(a, l, r)
    
    #use partition3
    lt, gt = partition3(a, l, r)

    randomized_quick_sort(a, l, lt);
    randomized_quick_sort(a, gt, r);
  
def main():
  input = sys.stdin.read()
  input = list(map(int, input.split()))
  n, a = input[0], input[1:]
  randomized_quick_sort(a, 0, n - 1)
  
  print " ".join([str(x) for x in a])

MAINFUNC = main
###########################################################################################
import traceback

if __name__ == '__main__':
  try:
    ret = MAINFUNC()
  except Exception as e:
    details = traceback.format_exc()
    print "\nRuntime error: %s\n\n %s"%(e, details)
  else:
    pass

File: Qsort_3w/src/Qsort_3w.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using std::vector;
using std::swap;

int partition2(vector<int> &a, int l, int r) {
  int x = a[l];
  int j = l;
  for (int i = l + 1; i <= r; i++) {
    if (a[i] <= x) {
      j++;
      swap(a[i], a[j]);
    }
  }
  swap(a[l], a[j]);
  return j;
}

//write down your code 
int partition3(vector<int> &a, int l, int r, int &lm, int &rm){
  int x = a[l];

  vector<int> am, al, ar;
  for (int i = l+1; i <= r; i++) {
    if (a[i] < x) al.push_back(a[i]);
    else if (a[i] == x) am.push_back(a[i]);
    else ar.push_back(a[i]); 
  }
  am.push_back(x);
  
  lm = l + (int)al.size();
  rm = lm + (int)am.size();
  
  al.insert(al.end(), am.begin(), am.end());
  al.insert(al.end(), ar.begin(), ar.end());
  al.insert(al.begin(), a.begin(), a.begin()+l);
  al.insert(al.end(), a.begin()+r+1, a.end());
  a.clear();
  a.assign(al.begin(), al.end());
  // std::cout<<std::endl;
  //for (size_t i = 0; i < a.size(); ++i) {
  //  std::cout << a[i] << ' ';
 // }
 // std::cout<<std::endl;

  return 0;
}


void randomized_quick_sort(vector<int> &a, int l, int r) {
  if (l >= r) {
    return;
  }

  int k = l + rand() % (r - l + 1);
  swap(a[l], a[k]);
  //int m = partition2(a, l, r);

  int lm, rm;
  //std::cout<<a[l]<<":";
  partition3(a, l, r, lm, rm);
  //std::cout<<l<<" "<<lm<<" "<<rm<<" "<<r<<std::endl;

  randomized_quick_sort(a, l, lm - 1);
  randomized_quick_sort(a, rm, r);
}

int main() {
  int n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < a.size(); ++i) {
    std::cin >> a[i];
  }
  randomized_quick_sort(a, 0, a.size() - 1);
  for (size_t i = 0; i < a.size(); ++i) {
    std::cout << a[i];
    if(i != a.size()-1) std::cout << ' ';
  }
  std::cout<<std::endl;
}
